module-system/module-implementations/sov-attester-incentives/src/call.rs

```diff
@@ -0,0 +1,764 @@
+
+        let bond = bond_opt.unwrap_or_default();
+        let bond: u64 = BorshDeserialize::deserialize(&mut bond.value())
+            .map_err(|_err| AttesterIncentiveErrors::InvalidBondFormat)?;
+
+        let minimum_bond = self
+            .minimum_attester_bond
+            .get_or_err(working_set)
+            .expect("The minimum bond should be set at genesis");
+
+        // We then have to check that the bond was greater than the minimum bond
+        if bond < minimum_bond {
+            return Err(AttesterIncentiveErrors::UserNotBonded);
+        }
+
+        Ok(())
+    }
+
+    fn check_transition(
+        &self,
+        max_attested_height: u64,
+        attester: &C::Address,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        if let Some(curr_tx) = self
+            .chain_state
+            .historical_transitions
+            .get(&max_attested_height, working_set)
+        {
+            // We first need to compare the initial block hash to the previous post state root
+            if !curr_tx.compare_hashes(&attestation.da_block_hash, &attestation.post_state_root) {
+                // It is the right attestation, we have to compare the initial block hash to the
+                // previous post state root
+                // Slash the attester
+                return Err(self.slash_and_invalidate_attestation(
+                    attester,
+                    max_attested_height,
+                    SlashingReason::TransitionInvalid,
+                    working_set,
+                ));
+            }
+            Ok(CallResponse::default())
+        } else {
+            Err(self.slash_burn_reward(
+                attester,
+                Role::Attester,
+                SlashingReason::TransitionNotFound,
+                working_set,
+            ))
+        }
+    }
+
+    fn check_initial_hash(
+        &self,
+        max_attested_height: u64,
+        attester: &C::Address,
+        attestation: &Attestation<StorageProof<<C::Storage as Storage>::Proof>>,
+        working_set: &mut WorkingSet<C::Storage>,
+    ) -> Result<sov_modules_api::CallResponse, AttesterIncentiveErrors> {
+        // Normal state
+        if let Some(transition) = self
+            .chain_state
+            .historical_transitions
+            .get(&(max_attested_height - 1), working_set)
+        {
+            if transition.post_state_root() != attestation.initial_state_root {
+                // The initial root hashes don't match, just slash the attester
+                return Err(self.slash_burn_reward(
+                    attester,
+                    Role::Attester,
+                    SlashingReason::InvalidInitialHash,
+                    working_set,
+                ));
+            }
```

Comment:
Since the param is `max_attested_height`, I think we end up slashing any attesters who attest to old transitions. But it's possible for multiple attestations to the same transition to land on-chain at the same time - so whoever gets ordered second by the DA layer will get slashed. 

I think we need to change this to work on the attestation's claimed `transition_num`, not on the `max_attested_height`. That way, we'll slash attesters who attest to an invalid root regardless of whether they're the first attestation for that transition, but we won't penalize people just for being the second to attest to a transition.

Reply:
Corrected with the last commit